File: Frontend/src/services/deliberationEngine.ts
export interface ECUEResult {
  code: string;
  nom: string;
  note: number; // /20
  coef: number;
}

export interface UEResult {
  code: string;
  nom: string;
  ects: number;
  ecues: ECUEResult[];
  moyenneUE: number;
  validee: boolean;
  valideeParCompensation: boolean;
  hasNoteEliminatoire: boolean;
}

export interface EtudiantDeliberation {
  id: string;
  matricule: string;
  nom: string;
  prenom: string;
  filiere: string;
  semestre: string;
  ues: UEResult[];
  moyenneGenerale: number;
  totalECTSAcquis: number;
  totalECTSMax: number;
  statutSession: "Admis" | "Rattrapage" | "Ajourné";
  mention: "Très Bien" | "Bien" | "Assez Bien" | "Passable" | "Ajourné" | "Rattrapage";
  hasNoteEliminatoire: boolean;
  notesEliminatoiresDetails: string[];
}

export interface DeliberationConfig {
  seuilValidationMoyenne: number; // Default 10.0
  seuilEliminatoire: number; // Seuil 7.0 : une note inférieure déclenche un rattrapage
  seuilPassageConditionnelECTS: number; // Default 18 ECTS out of 30
  compensationAutorisee: boolean; // Default true (inter-UE compensation if sem average >= 10)
}

export const DEFAULT_DELIBERATION_CONFIG: DeliberationConfig = {
  seuilValidationMoyenne: 10.0,
  seuilEliminatoire: 7.0,
  seuilPassageConditionnelECTS: 18,
  compensationAutorisee: true,
};

export interface DeliberationStudentInput {
  id: string;
  matricule: string;
  nom: string;
  prenom: string;
  filiere: string;
  semestre: string;
  uesRaw: {
    code: string;
    nom: string;
    ects: number;
    ecues: ECUEResult[];
  }[];
}

export const deliberationEngine = {
  // Calculer la délibération individuelle pour un étudiant
  calculerEtudiant: (
    etudiantRaw: DeliberationStudentInput,
    config: DeliberationConfig = DEFAULT_DELIBERATION_CONFIG
  ): EtudiantDeliberation => {
    let sumECTSValides = 0;
    let sumECTSWeighted = 0;
    let sumECTSTotal = 0;
    let globalHasEliminatoire = false;
    const allEliminatoires: string[] = [];

    // 1. Calcul des moyennes par UE
    const uesCalculees: UEResult[] = etudiantRaw.uesRaw.map((ue) => {
      const totalNoteCoef = ue.ecues.reduce((sum, e) => sum + e.note * e.coef, 0);
      const totalCoef = ue.ecues.reduce((sum, e) => sum + e.coef, 0);
      const moyenneUE = totalCoef > 0 ? Number((totalNoteCoef / totalCoef).toFixed(2)) : 0;

      // Check note éliminatoire dans l'UE
      const ueEliminatoires = ue.ecues.filter((e) => e.note < config.seuilEliminatoire);
      const hasNoteEliminatoire = ueEliminatoires.length > 0;

      if (hasNoteEliminatoire) {
        globalHasEliminatoire = true;
        ueEliminatoires.forEach((e) => {
          allEliminatoires.push(`${e.code} (${e.note}/20)`);
        });
      }

      sumECTSTotal += ue.ects;
      sumECTSWeighted += moyenneUE * ue.ects;

      return {
        code: ue.code,
        nom: ue.nom,
        ects: ue.ects,
        ecues: ue.ecues,
        moyenneUE,
        validee: moyenneUE >= config.seuilValidationMoyenne && !hasNoteEliminatoire,
        valideeParCompensation: false,
        hasNoteEliminatoire,
      };
    });

    // 2. Moyenne générale du semestre
    const moyenneGenerale =
      sumECTSTotal > 0 ? Number((sumECTSWeighted / sumECTSTotal).toFixed(2)) : 0;

    // 3. Application des règles de compensation
    const canCompensate =
      config.compensationAutorisee &&
      moyenneGenerale >= config.seuilValidationMoyenne &&
      !globalHasEliminatoire;

    uesCalculees.forEach((ue) => {
      if (ue.validee) {
        sumECTSValides += ue.ects;
      } else if (canCompensate && !ue.hasNoteEliminatoire) {
        ue.validee = true;
        ue.valideeParCompensation = true;
        sumECTSValides += ue.ects;
      }
    });

    // 4. Détermination automatique de la décision du jury
    let statutSession: EtudiantDeliberation["statutSession"] = "Ajourné";
    let mention: EtudiantDeliberation["mention"] = "Ajourné";

    if (moyenneGenerale >= config.seuilValidationMoyenne && sumECTSValides === sumECTSTotal) {
      statutSession = "Admis";
      if (moyenneGenerale >= 16) mention = "Très Bien";
      else if (moyenneGenerale >= 14) mention = "Bien";
      else if (moyenneGenerale >= 12) mention = "Assez Bien";
      else mention = "Passable";
    } else if (
      sumECTSValides >= config.seuilPassageConditionnelECTS ||
      globalHasEliminatoire ||
      (moyenneGenerale >= 8.5 && moyenneGenerale < config.seuilValidationMoyenne)
    ) {
      statutSession = "Rattrapage";
      mention = "Rattrapage";
    } else {
      statutSession = "Ajourné";
      mention = "Ajourné";
    }

    return {
      id: etudiantRaw.id,
      matricule: etudiantRaw.matricule,
      nom: etudiantRaw.nom,
      prenom: etudiantRaw.prenom,
      filiere: etudiantRaw.filiere,
      semestre: etudiantRaw.semestre,
      ues: uesCalculees,
      moyenneGenerale,
      totalECTSAcquis: sumECTSValides,
      totalECTSMax: sumECTSTotal,
      statutSession,
      mention,
      hasNoteEliminatoire: globalHasEliminatoire,
      notesEliminatoiresDetails: allEliminatoires,
    };
  },

  // Calculer toute une cohorte / promotion
  calculerPromotion: (
    etudiants: DeliberationStudentInput[],
    config: DeliberationConfig = DEFAULT_DELIBERATION_CONFIG
  ): {
    resultats: EtudiantDeliberation[];
    stats: {
      total: number;
      admis: number;
      rattrapage: number;
      ajournes: number;
      tauxReussite: string;
      moyennePromo: string;
    };
  } => {
    const resultats = etudiants.map((e) => deliberationEngine.calculerEtudiant(e, config));

    const total = resultats.length;
    const admis = resultats.filter((r) => r.statutSession === "Admis").length;
    const rattrapage = resultats.filter((r) => r.statutSession === "Rattrapage").length;
    const ajournes = resultats.filter((r) => r.statutSession === "Ajourné").length;

    const sumMoyennes = resultats.reduce((acc, r) => acc + r.moyenneGenerale, 0);
    const moyennePromo = total > 0 ? (sumMoyennes / total).toFixed(2) : "0.00";
    const tauxReussite = total > 0 ? `${((admis / total) * 100).toFixed(1)}%` : "0%";

    return {
      resultats,
      stats: {
        total,
        admis,
        rattrapage,
        ajournes,
        tauxReussite,
        moyennePromo,
      },
    };
  },
};

File: Backend/app/services/deliberation_engine.py
"""
Moteur de Délibération ECTS / LMD pour EduManagePro (EMP).
Calcule les moyennes d'UEs, applique la compensation, détermine les statuts
(Admis, Rattrapage, Ajourné) et les mentions académiques.
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class ECUEResult(BaseModel):
    code: str
    nom: str
    note: float  # /20
    coef: float = 1.0


class UEResult(BaseModel):
    code: str
    nom: str
    ects: int
    ecues: List[ECUEResult]
    moyenne_ue: float
    validee: bool
    validee_par_compensation: bool
    has_note_eliminatoire: bool


class EtudiantDeliberationResult(BaseModel):
    id: Optional[str] = None
    matricule: str
    nom: str
    prenom: str
    filiere: str
    semestre: str
    ues: List[UEResult]
    moyenne_generale: float
    total_ects_acquis: int
    total_ects_max: int
    statut_session: str  # "Admis" | "Rattrapage" | "Ajourné"
    mention: str         # "Très Bien" | "Bien" | "Assez Bien" | "Passable" | "Rattrapage" | "Ajourné"
    has_note_eliminatoire: bool
    notes_eliminatoires_details: List[str]


class DeliberationConfig(BaseModel):
    seuil_validation_moyenne: float = 10.0
    seuil_eliminatoire: float = 7.0
    seuil_passage_conditionnel_ects: int = 18
    compensation_autorisee: bool = True


class PromotionDeliberationStats(BaseModel):
    total: int
    admis: int
    rattrapage: int
    ajournes: int
    taux_reussite: str
    moyenne_promo: str


class PromotionDeliberationResult(BaseModel):
    resultats: List[EtudiantDeliberationResult]
    stats: PromotionDeliberationStats


class DeliberationEngine:
    @staticmethod
    def calculer_etudiant(
        etudiant_data: Dict[str, Any],
        config: DeliberationConfig = DeliberationConfig()
    ) -> EtudiantDeliberationResult:
        """
        Calcule les résultats individuels de délibération pour un étudiant.
        """
        sum_ects_valides = 0
        sum_ects_weighted = 0.0
        sum_ects_total = 0
        global_has_eliminatoire = False
        all_eliminatoires: List[str] = []

        ues_calculees: List[UEResult] = []

        # 1. Calcul par UE
        ues_raw = etudiant_data.get("ues_raw", [])
        for ue in ues_raw:
            ecues_raw = ue.get("ecues", [])
            total_note_coef = sum(e["note"] * e.get("coef", 1.0) for e in ecues_raw)
            total_coef = sum(e.get("coef", 1.0) for e in ecues_raw)
            moyenne_ue = round(total_note_coef / total_coef, 2) if total_coef > 0 else 0.0

            # Vérification des notes éliminatoires dans l'UE
            ue_eliminatoires = [e for e in ecues_raw if e["note"] < config.seuil_eliminatoire]
            has_ue_eliminatoire = len(ue_eliminatoires) > 0

            if has_ue_eliminatoire:
                global_has_eliminatoire = True
                for e in ue_eliminatoires:
                    all_eliminatoires.append(f"{e.get('code', 'EC')} ({e['note']}/20)")

            ects = ue.get("ects", 0)
            sum_ects_total += ects
            sum_ects_weighted += moyenne_ue * ects

            is_validee = (moyenne_ue >= config.seuil_validation_moyenne) and (not has_ue_eliminatoire)

            ues_calculees.append(
                UEResult(
                    code=ue.get("code", ""),
                    nom=ue.get("nom", ""),
                    ects=ects,
                    ecues=[ECUEResult(**e) for e in ecues_raw],
                    moyenne_ue=moyenne_ue,
                    validee=is_validee,
                    validee_par_compensation=False,
                    has_note_eliminatoire=has_ue_eliminatoire,
                )
            )

        # 2. Moyenne générale
        moyenne_generale = (
            round(sum_ects_weighted / sum_ects_total, 2) if sum_ects_total > 0 else 0.0
        )

        # 3. Règles de compensation
        can_compensate = (
            config.compensation_autorisee
            and moyenne_generale >= config.seuil_validation_moyenne
            and not global_has_eliminatoire
        )

        for ue_calc in ues_calculees:
            if ue_calc.validee:
                sum_ects_valides += ue_calc.ects
            elif can_compensate and not ue_calc.has_note_eliminatoire:
                ue_calc.validee = True
                ue_calc.validee_par_compensation = True
                sum_ects_valides += ue_calc.ects

        # 4. Décision du Jury & Mention
        if moyenne_generale >= config.seuil_validation_moyenne and sum_ects_valides == sum_ects_total:
            statut_session = "Admis"
            if moyenne_generale >= 16.0:
                mention = "Très Bien"
            elif moyenne_generale >= 14.0:
                mention = "Bien"
            elif moyenne_generale >= 12.0:
                mention = "Assez Bien"
            else:
                mention = "Passable"
        elif (
            sum_ects_valides >= config.seuil_passage_conditionnel_ects
            or global_has_eliminatoire
            or (8.5 <= moyenne_generale < config.seuil_validation_moyenne)
        ):
            statut_session = "Rattrapage"
            mention = "Rattrapage"
        else:
            statut_session = "Ajourné"
            mention = "Ajourné"

        return EtudiantDeliberationResult(
            id=etudiant_data.get("id"),
            matricule=etudiant_data.get("matricule", ""),
            nom=etudiant_data.get("nom", ""),
            prenom=etudiant_data.get("prenom", ""),
            filiere=etudiant_data.get("filiere", ""),
            semestre=etudiant_data.get("semestre", "Semestre 1"),
            ues=ues_calculees,
            moyenne_generale=moyenne_generale,
            total_ects_acquis=sum_ects_valides,
            total_ects_max=sum_ects_total,
            statut_session=statut_session,
            mention=mention,
            has_note_eliminatoire=global_has_eliminatoire,
            notes_eliminatoires_details=all_eliminatoires,
        )

    @classmethod
    def calculer_promotion(
        cls,
        etudiants: List[Dict[str, Any]],
        config: DeliberationConfig = DeliberationConfig()
    ) -> PromotionDeliberationResult:
        """
        Calcule les résultats pour l'ensemble d'une promotion/cohorte.
        """
        resultats = [cls.calculer_etudiant(e, config) for e in etudiants]
        total = len(resultats)
        admis = sum(1 for r in resultats if r.statut_session == "Admis")
        rattrapage = sum(1 for r in resultats if r.statut_session == "Rattrapage")
        ajournes = sum(1 for r in resultats if r.statut_session == "Ajourné")

        sum_moyennes = sum(r.moyenne_generale for r in resultats)
        moyenne_promo = f"{(sum_moyennes / total):.2f}" if total > 0 else "0.00"
        taux_reussite = f"{((admis / total) * 100):.1f}%" if total > 0 else "0.0%"

        return PromotionDeliberationResult(
            resultats=resultats,
            stats=PromotionDeliberationStats(
                total=total,
                admis=admis,
                rattrapage=rattrapage,
                ajournes=ajournes,
                taux_reussite=taux_reussite,
                moyenne_promo=moyenne_promo,
            )
        )
